### PR TITLE
build(wasm-builder): improve toolchain detection with regex

### PR DESCRIPTION
**Pull request by DennisInSky**
@reviewer-or-team

### CODE REVIEW

File: utils/wasm-builder/src/cargo_toolchain.rs [hunk 1]
```diff
@@ -43,9 +43,12 @@ impl Toolchain {
             .ok_or_else(|| BuilderError::CargoPathInvalid(path.clone()))?;
 
         // Toolchain name format:
-        // "**toolchain**-arch-arch-arch-arch"
+        // "**toolchain**-arch-arch-arch-arch" (linux/windows?)
```
Comment (shamilsan) [inline utils/wasm-builder/src/cargo_toolchain.rs:46]: Here are all targets. Maybe we are to find more robust solution?

```
aarch64-apple-darwin
aarch64-apple-ios
aarch64-apple-ios-sim
aarch64-linux-android
aarch64-pc-windows-msvc
aarch64-unknown-fuchsia
aarch64-unknown-linux-gnu
aarch64-unknown-linux-musl
aarch64-unknown-none
aarch64-unknown-none-softfloat
aarch64-unknown-uefi
arm-linux-androideabi
arm-unknown-linux-gnueabi
arm-unknown-linux-gnueabihf
arm-unknown-linux-musleabi
arm-unknown-linux-musleabihf
armebv7r-none-eabi
armebv7r-none-eabihf
armv5te-unknown-linux-gnueabi
armv5te-unknown-linux-musleabi
armv7-linux-androideabi
armv7-unknown-linux-gnueabi
armv7-unknown-linux-gnueabihf
armv7-unknown-linux-musleabi
armv7-unknown-linux-musleabihf
armv7a-none-eabi
armv7r-none-eabi
armv7r-none-eabihf
asmjs-unknown-emscripten
i586-pc-windows-msvc
i586-unknown-linux-gnu
i586-unknown-linux-musl
i686-linux-android
i686-pc-windows-gnu
i686-pc-windows-msvc
i686-unknown-freebsd
i686-unknown-linux-gnu
i686-unknown-linux-musl
i686-unknown-uefi
mips-unknown-linux-gnu
mips-unknown-linux-musl
mips64-unknown-linux-gnuabi64
mips64-unknown-linux-muslabi64
mips64el-unknown-linux-gnuabi64
mips64el-unknown-linux-muslabi64
mipsel-unknown-linux-gnu
mipsel-unknown-linux-musl
nvptx64-nvidia-cuda
powerpc-unknown-linux-gnu
powerpc64-unknown-linux-gnu
powerpc64le-unknown-linux-gnu
riscv32i-unknown-none-elf
riscv32imac-unknown-none-elf
riscv32imc-unknown-none-elf
riscv64gc-unknown-linux-gnu
riscv64gc-unknown-none-elf
riscv64imac-unknown-none-elf
s390x-unknown-linux-gnu
sparc64-unknown-linux-gnu
sparcv9-sun-solaris
thumbv6m-none-eabi
thumbv7em-none-eabi
thumbv7em-none-eabihf
thumbv7m-none-eabi
thumbv7neon-linux-androideabi
thumbv7neon-unknown-linux-gnueabihf
thumbv8m.base-none-eabi
thumbv8m.main-none-eabi
thumbv8m.main-none-eabihf
wasm32-unknown-emscripten
wasm32-unknown-unknown (installed)
wasm32-wasi
x86_64-apple-darwin
x86_64-apple-ios
x86_64-fortanix-unknown-sgx
x86_64-linux-android
x86_64-pc-solaris
x86_64-pc-windows-gnu
x86_64-pc-windows-msvc
x86_64-sun-solaris
x86_64-unknown-freebsd
x86_64-unknown-fuchsia
x86_64-unknown-illumos
x86_64-unknown-linux-gnu
x86_64-unknown-linux-gnux32
x86_64-unknown-linux-musl
x86_64-unknown-netbsd
x86_64-unknown-none
x86_64-unknown-redox
x86_64-unknown-uefi
```

**Review by shamilsan — APPROVED** [review]
Ok as a quick and dirty fix, but maybe we are to use regex or even better solution.

**Comment by DennisInSky** [comment]
Fixed using @StackOverflowExcept1on suggestion, thanks to him

**Comment by StackOverflowExcept1on** [comment]
CI passed. Please make merge with co-authors

File: utils/wasm-builder/src/cargo_toolchain.rs [hunk 2]
```diff
@@ -35,20 +46,30 @@ impl Toolchain {
...
 
-        // Toolchain name format:
-        // "**toolchain**-arch-arch-arch-arch"
-        let toolchain = toolchain_name
-            .rsplitn(5, '-')
-            .last()
-            .map(String::from)
-            .ok_or_else(|| BuilderError::CargoToolchainInvalid(toolchain_name.into()))?;
+        // This regex is borrowed from the rustup code and modified (added non-capturing groups)
+        lazy_static! {
```
Comment (shamilsan) [inline utils/wasm-builder/src/cargo_toolchain.rs:57]: Not sure `lazy_static` is good inside the function, but no matters. And I personally prefer `once_cell` as it seems to be in `std` lib in some future 😄 
https://github.com/rust-lang/rust/issues/74465

**Review by shamilsan — APPROVED** [review]

File: utils/wasm-builder/src/cargo_toolchain.rs (hunk 2 above)
Comment (DennisInSky) [reply utils/wasm-builder/src/cargo_toolchain.rs]: Re-implemented via once_cell

**Pull request merged**